### PR TITLE
Epm UI i 7279 personal data modal

### PR DESCRIPTION
**Pull request by AlexanderKrasovskiy**
1. Task ([EPMUII-7279](https://jira.epam.com/jira/browse/EPMUII-7279)): Add Modal with Contact Info form
2. Screenshots:
modal_FINAL_0
modal_FINAL

### CODE REVIEW

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        color="primary"
+        [mat-dialog-close]="false"
+        *ngIf="data.source === openedFrom.RegistrationPage"
+      >
+        {{ 'пропустить' | uppercase }}
+      </button>
+      <button
+        class="btn btn-cancel"
+        mat-raised-button
+        color="primary"
```
Comment (vyMoon) [inline src/app/shared/components/personal-info-modal/personal-info-modal.component.html:83]: if you change this attr the color is changed.

File: src/app/auth/services/api/api.service.ts [hunk 2]
```diff
@@ -26,4 +27,13 @@ export class ApiService {
...
     return this.http.post<any>('password_reset', data);
   }
+
+  addContactInfo({ name, surname, phoneNumber }: UserPersonalInfo) {
+    const data = {
+      name,
+      surname,
+      phone: phoneNumber,
+    };
+    return this.http.patch<any>('v1/users/me', data);
```
Comment (vyMoon) [inline src/app/auth/services/api/api.service.ts:37]: never use any

File: src/app/auth/components/sign-up/sign-up.component.ts [hunk 3]
```diff
@@ -67,14 +73,19 @@ export class SignUpComponent implements OnInit {
...
         catchError((error) => {
           this.notificationsService.openError(error.message);
           return throwError(error);
         }),
         finalize(() => this.blockUiService.unBlock()),
-        take(1),
+        untilDestroyed(this),
       )
-      .subscribe(() => this.router.navigate(['/']));
+      .subscribe();
```
Comment (vyMoon) [inline src/app/auth/components/sign-up/sign-up.component.ts:88]: never leave subscribe empty

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+</div>
+
+<div class="message-container">
+  <p class="message-text" *ngIf="data.source === openedFrom.OrderPage">
+    Для отправки заявки укажите Ваши персональные данные
+  </p>
+</div>
+
+<mat-tab-group mat-align-tabs="center">
+  <mat-tab label="{{ 'укажите ваши персональные данные' | uppercase }}">
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/components/personal-info-modal/personal-info-modal.component.html:12]: done

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      </button>
+      <button
+        class="btn btn-cancel"
+        mat-raised-button
+        color="primary"
+        [mat-dialog-close]="false"
+        *ngIf="data.source === openedFrom.OrderPage"
+      >
+        {{ 'отменить' | uppercase }}
+      </button>
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/components/personal-info-modal/personal-info-modal.component.html:88]: done

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      </mat-form-field>
+
+      <mat-form-field floatLabel="auto">
+        <mat-label>Телефон</mat-label>
+        <input
+          type="tel"
+          name="phoneNumber"
+          matInput
+          minlength="9"
+          maxlength="20"
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/components/personal-info-modal/personal-info-modal.component.html:49]: done - changed to 11 - typical mobile number length

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.scss [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    color: $color-title;
+  }
+}
+
+.btn {
+  width: 100%;
+  margin-top: 10px;
+  &-cancel {
+    background-color: #a9a9a9;
+  }
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/components/personal-info-modal/personal-info-modal.component.scss:55]: Unfortunately, I didn't find this color in our custom palette. As far as I understand, (according to Figma) this color is solely used for "Cancel" button.

File: src/app/shared/services/personal-info/personal-info.service.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        maxWidth: '100vw',
+        position: { top: '0' },
+        data,
+      })
+      .afterClosed()
+      .pipe(
+        filter(Boolean),
+        switchMap((data) => this.api.addContactInfo(data)),
+        catchError((error) => {
+          this.notificationsService.openError(error.message);
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/services/personal-info/personal-info.service.ts:41]: done - removed

File: src/app/auth/services/api/api.service.ts [hunk 9]
```diff
@@ -26,4 +27,13 @@ export class ApiService {
     const data = { data: { login: email } };
     return this.http.post<any>('password_reset', data);
   }
+
+  addContactInfo({ name, surname, phoneNumber }: UserPersonalInfo) {
+    const data = {
+      name,
+      surname,
+      phone: phoneNumber,
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/auth/services/api/api.service.ts:35]: done - replaced phoneNuber with number

File: src/app/shared/services/personal-info/personal-info.service.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    private readonly api: ApiService,
+    private readonly notificationsService: NotificationsService,
+  ) {}
+
+  openPersonalInfoModal(source: PersonalInfo) {
+    const data: UserPersonalInfo = {
+      name: '',
+      surname: '',
+      phoneNumber: '',
+      source,
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/services/personal-info/personal-info.service.ts:25]: Our app is expected to open this modal from several places - after registration and from order page. Modal's contents change a bit depending on the source (from registration or from order page). That's why we need this info to change contents accordingly.

File: src/app/shared/constants/personal-info.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import { PersonalInfo } from '@shared/constants/personal-info.enum';
+
+export interface UserPersonalInfo {
+  name: string;
+  surname: string;
+  phoneNumber: string;
+  source: PersonalInfo;
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/constants/personal-info.ts:7]: please, see the previous answer

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+@Component({
+  selector: 'lc-personal-info-modal',
+  templateUrl: './personal-info-modal.component.html',
+  styleUrls: ['./personal-info-modal.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class PersonalInfoModalComponent {
+  readonly openedFrom = PersonalInfo;
+
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/components/personal-info-modal/personal-info-modal.component.ts:14]: done

File: src/app/shared/services/personal-info/personal-info.service.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    };
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/services/personal-info/personal-info.service.ts:26]: done

File: src/app/shared/constants/notification-success.enum.ts [hunk 14]
```diff
@@ -1,4 +1,5 @@
 export enum NotificationSuccess {
   Authorized = 'Авторизация прошла упешно.',
   EquipmentAdded = 'Оборудование зарегистрировано успешно',
+  ContactInfoAdded = 'Вы успешно зарегистрировались, ваши данные для входа отправлены на почту',
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/constants/notification-success.enum.ts:4]: This message was shown in Figma. Should I remove this text in enum? Additionally removed opening this toast after submit registration.

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html (hunk 1 above)
Comment (AlexanderKrasovskiy) [reply src/app/shared/components/personal-info-modal/personal-info-modal.component.html:83]: unfortunately we did not specify this color in our custom theme. As far as i see from Figma, this color occurs only once in Cancel button

File: src/app/auth/services/api/api.service.ts (hunk 2 above)
Comment (AlexanderKrasovskiy) [reply src/app/auth/services/api/api.service.ts:37]: done - changed to void

File: src/app/auth/components/sign-up/sign-up.component.ts (hunk 3 above)
Comment (AlexanderKrasovskiy) [reply src/app/auth/components/sign-up/sign-up.component.ts:88]: done

File: src/app/shared/services/validation/validation.service.ts [hunk 15]
```diff
@@ -14,6 +14,7 @@ export class ValidationService {
     maxlength: 'Слишком большое знанеие, уменьшите количество символов',
     minlength: 'Слишком короткое значение, добавтесимволы',
     email: 'Значение должно быть типа: example@mail.com',
+    pattern: 'Номер телефона должен состоять исключительно из цифр',
```
[the remark this answers is not in the document]
Comment (AlexanderKrasovskiy) [reply src/app/shared/services/validation/validation.service.ts:17]: done - add separate PhoneValidator directive

File: src/app/auth/components/sign-up/sign-up.component.spec.ts [hunk 16]
```diff
@@ -1,14 +1,61 @@
...
+const validationServiceMock = jasmine.createSpyObj('ValidationService', {
+  emitSubmit: undefined,
+  compare: () => null,
+});
+
+const blockUiServiceMock = jasmine.createSpyObj('BlockUiService', ['block', 'unBlock']);
+
+const notificationsServiceMock = jasmine.createSpyObj('NotificationsService', ['openError']);
+
 xdescribe('SignUpComponent', () => {
```
Comment (denezhkinia) [inline src/app/auth/components/sign-up/sign-up.component.spec.ts:30]: delete x for start work test

File: src/app/shared/directives/phone-validator/phone-validator.directive.spec.ts [hunk 17]
```diff
@@ -0,0 +1,8 @@
+import { PhoneValidatorDirective } from './phone-validator.directive';
+
+xdescribe('PhoneValidatorDirective', () => {
+  it('should create an instance', () => {
+    const directive = new PhoneValidatorDirective();
+    expect(directive).toBeTruthy();
+  });
+});
```
Comment (denezhkinia) [inline src/app/shared/directives/phone-validator/phone-validator.directive.spec.ts:8]: add tests

File: src/app/auth/components/sign-up/sign-up.component.spec.ts (hunk 16 above)
Comment (denezhkinia) [reply src/app/auth/components/sign-up/sign-up.component.spec.ts:30]: and add tests

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+          required
+          [(ngModel)]="contacts.phone"
+          lcPhoneValidator
+          lcValidationErrors
+        />
+      </mat-form-field>
+
+      <div class="offer">
+        Нажимая «Продолжить» вы даете
+        <a routerLink="public-offer" target="_blank">согласие на обработку персональных данных</a>
```
Comment (vyMoon) [inline src/app/shared/components/personal-info-modal/personal-info-modal.component.html:59]: click on public offer leads to 404

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+        <input
+          type="tel"
+          name="phone"
+          matInput
+          minlength="11"
+          maxlength="11"
+          required
+          [(ngModel)]="contacts.phone"
+          lcPhoneValidator
+          lcValidationErrors
```
Comment (vyMoon) [inline src/app/shared/components/personal-info-modal/personal-info-modal.component.html:53]: the fields displayed with validation errors

File: src/app/shared/services/personal-info/personal-info.service.ts [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+import { ApiService } from '@app/auth/services/api/api.service';
+import { UserPersonalInfo } from '@app/shared/constants/personal-info';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class PersonalInfoService {
+  constructor(private readonly dialog: MatDialog, private readonly api: ApiService) {}
+
+  openPersonalInfoModal(source: OpenedFrom) {
```
Comment (vyMoon) [inline src/app/shared/services/personal-info/personal-info.service.ts:15]: returned type is missing

File: src/app/shared/services/personal-info/personal-info.service.ts [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+        position: { top: '0' },
+        data: source,
+      })
+      .afterClosed()
+      .pipe(
+        switchMap((contactInfo: UserPersonalInfo | false) => {
+          if (contactInfo) {
+            return this.api.addContactInfo(contactInfo);
+          } else {
+            return of('');
```
Comment (vyMoon) [inline src/app/shared/services/personal-info/personal-info.service.ts:30]: let's return undefined

File: src/app/auth/services/api/api.service.ts [hunk 22]
```diff
@@ -26,4 +27,8 @@ export class ApiService {
     const data = { data: { login: email } };
     return this.http.post<any>('password_reset', data);
   }
+
+  addContactInfo(data: UserPersonalInfo) {
```
Comment (vyMoon) [inline src/app/auth/services/api/api.service.ts:31]: returned type is missing

File: src/app/auth/services/controller/controller.service.ts [hunk 23]
```diff
@@ -53,4 +56,8 @@ export class ControllerService {
   setRememberMe(rememberMe: boolean) {
     this.store.dispatch(new rememberMeAction(rememberMe));
   }
+
+  openPersonalInfoModal(source: OpenedFrom) {
```
Comment (vyMoon) [inline src/app/auth/services/controller/controller.service.ts:60]: returned type is missing

File: src/app/shared/directives/phone-validator/phone-validator.directive.ts [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+  providers: [{ provide: NG_VALIDATORS, useExisting: PhoneValidatorDirective, multi: true }],
+})
+export class PhoneValidatorDirective implements Validator {
+  validate(control: AbstractControl): ValidationErrors | null {
+    const PHONE_REGEXP = /[0-9]+/;
+    const isValid = PHONE_REGEXP.test(control.value);
+
+    return isValid ? null : { phoneValidator: { message: 'Номер телефона должен состоять исключительно из цифр' } };
+  }
+}
```
Comment (vyMoon) [inline src/app/shared/directives/phone-validator/phone-validator.directive.ts:15]: I think we do not need this. I would create reactive form and set custom validator for such a purpose.

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html (hunk 18 above)
Comment (AlexanderKrasovskiy) [reply src/app/shared/components/personal-info-modal/personal-info-modal.component.html:59]: fixed link, but AuthGuard redirects to catalog. We need to discuss other route for public offer

File: src/app/shared/services/personal-info/personal-info.service.ts (hunk 20 above)
Comment (AlexanderKrasovskiy) [reply src/app/shared/services/personal-info/personal-info.service.ts:15]: done

File: src/app/shared/services/personal-info/personal-info.service.ts (hunk 21 above)
Comment (AlexanderKrasovskiy) [reply src/app/shared/services/personal-info/personal-info.service.ts:30]: done

File: src/app/auth/services/api/api.service.ts (hunk 22 above)
Comment (AlexanderKrasovskiy) [reply src/app/auth/services/api/api.service.ts:31]: done

File: src/app/auth/services/controller/controller.service.ts (hunk 23 above)
Comment (AlexanderKrasovskiy) [reply src/app/auth/services/controller/controller.service.ts:60]: done

File: src/app/shared/directives/phone-validator/phone-validator.directive.spec.ts (hunk 17 above)
Comment (AlexanderKrasovskiy) [reply src/app/shared/directives/phone-validator/phone-validator.directive.spec.ts]: directive deleted

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html (hunk 18 above)
Comment (AlexanderKrasovskiy) [reply src/app/shared/components/personal-info-modal/personal-info-modal.component.html]: done as discussed

File: src/app/shared/components/personal-info-modal/personal-info-modal.component.html (hunk 19 above)
Comment (AlexanderKrasovskiy) [reply src/app/shared/components/personal-info-modal/personal-info-modal.component.html]: done

File: src/app/shared/directives/phone-validator/phone-validator.directive.ts (hunk 24 above)
Comment (AlexanderKrasovskiy) [reply src/app/shared/directives/phone-validator/phone-validator.directive.ts]: done - replaced to reactive form

**Pull request merged**